File: sdk/java/gladys-sensor-sdk/src/main/java/com/gladys/sensor/HeartbeatManager.java
package com.gladys.sensor;

import gladys.v1.Common;
import gladys.v1.Orchestrator;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Sends periodic heartbeats to the orchestrator on a background thread.
 * Optionally dispatches pending commands received in heartbeat responses.
 */
public class HeartbeatManager {

    private static final Logger logger = Logger.getLogger(HeartbeatManager.class.getName());

    private final GladysClient client;
    private final String componentId;
    private final int intervalSeconds;
    private final ScheduledExecutorService scheduler;
    private final AtomicBoolean running = new AtomicBoolean(false);
    private final Consumer<Orchestrator.PendingCommand> onCommand;
    private volatile Common.ComponentState state = Common.ComponentState.COMPONENT_STATE_ACTIVE;
    private volatile String errorMessage;

    public HeartbeatManager(GladysClient client, String componentId, int intervalSeconds) {
        this(client, componentId, intervalSeconds, null);
    }

    public HeartbeatManager(GladysClient client, String componentId, int intervalSeconds,
                            Consumer<Orchestrator.PendingCommand> onCommand) {
        this.client = client;
        this.componentId = componentId;
        this.intervalSeconds = intervalSeconds;
        this.onCommand = onCommand;
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "gladys-heartbeat-" + componentId);
            t.setDaemon(true);
            return t;
        });
    }

    public void start() {
        if (running.compareAndSet(false, true)) {
            scheduler.scheduleAtFixedRate(this::sendHeartbeat, 0, intervalSeconds, TimeUnit.SECONDS);
        }
    }

    public void stop() {
        if (running.compareAndSet(true, false)) {
            scheduler.shutdown();
            try {
                if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                    scheduler.shutdownNow();
                }
            } catch (InterruptedException e) {
                scheduler.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
    }

    public boolean isRunning() {
        return running.get();
    }

    public void setState(Common.ComponentState state) {
        this.state = state;
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    private void sendHeartbeat() {
        try {
            Orchestrator.HeartbeatResponse response = client.heartbeat(componentId, state, errorMessage);

            if (onCommand != null && response.getAcknowledged()
                    && response.getPendingCommandsCount() > 0) {
                for (Orchestrator.PendingCommand cmd : response.getPendingCommandsList()) {
                    try {
                        onCommand.accept(cmd);
                    } catch (Exception e) {
                        logger.log(Level.WARNING,
                                "Command callback failed for " + cmd.getCommand(), e);
                    }
                }
            }
        } catch (Exception e) {
            logger.log(Level.WARNING, "Heartbeat failed for " + componentId, e);
        }
    }
}

File: sdk/js/gladys-sensor-sdk/src/HeartbeatManager.ts
import { GladysClient } from "./GladysClient";
import { ComponentState } from "./generated/common";
import { type PendingCommand } from "./generated/orchestrator";

/**
 * Callback invoked for each pending command received in heartbeat response.
 */
export type OnCommandCallback = (cmd: PendingCommand) => Promise<void>;

/**
 * Background heartbeat manager using setInterval.
 * Timer is unref'd so it doesn't prevent Node.js process exit.
 * Default state is COMPONENT_STATE_ACTIVE; call setState() to change.
 *
 * Optionally accepts an onCommand callback that is invoked for each
 * pending command in the heartbeat response.
 */
export class HeartbeatManager {
  private readonly client: GladysClient;
  private readonly componentId: string;
  private readonly intervalMs: number;
  private readonly onCommand: OnCommandCallback | null;
  private timer: ReturnType<typeof setInterval> | null = null;
  private running: boolean = false;
  private currentState: ComponentState = ComponentState.COMPONENT_STATE_ACTIVE;
  private errorMessage: string = "";

  constructor(
    client: GladysClient,
    componentId: string,
    intervalSeconds: number,
    onCommand?: OnCommandCallback
  ) {
    this.client = client;
    this.componentId = componentId;
    this.intervalMs = intervalSeconds * 1000;
    this.onCommand = onCommand ?? null;
  }

  start(): void {
    if (this.running) return;
    this.running = true;
    this.sendHeartbeat();
    this.timer = setInterval(() => this.sendHeartbeat(), this.intervalMs);
    this.timer.unref();
  }

  stop(): void {
    if (!this.running) return;
    this.running = false;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  isRunning(): boolean {
    return this.running;
  }

  setState(state: ComponentState): void {
    this.currentState = state;
  }

  getState(): ComponentState {
    return this.currentState;
  }

  /**
   * Set error message to include in next heartbeat.
   * Cleared after a successful heartbeat send.
   */
  setErrorMessage(message: string): void {
    this.errorMessage = message;
  }

  private sendHeartbeat(): void {
    const errorMsg = this.errorMessage;

    this.client
      .heartbeat(this.componentId, this.currentState, errorMsg || undefined)
      .then((response) => {
        this.errorMessage = "";
        if (this.onCommand && response.pendingCommands?.length > 0) {
          // Process pending commands sequentially
          this.processCommands(response.pendingCommands);
        }
      })
      .catch(() => {
        // Swallow errors -- heartbeat failures shouldn't crash the sensor.
        // Orchestrator monitors heartbeat absence for dead sensor detection.
      });
  }

  private async processCommands(commands: PendingCommand[]): Promise<void> {
    for (const cmd of commands) {
      try {
        await this.onCommand!(cmd);
      } catch {
        // Command processing errors are handled by CommandDispatcher.
        // Don't let one command failure prevent processing the rest.
      }
    }
  }
}
